### PR TITLE
[HNT-413] ETL for HNT Content reporting

### PR DESCRIPTION
**Pull request by gkatre**
## Description

This etl will be used for HNT content based reporting. 

The initial use will be for the creating an equivalent of Mode report “[FX New Tab - Feed Engagement](https://app.mode.com/getpocket/reports/1d70145e0015/runs/931e35fe9928)” using Glean data

## Related Tickets & Documents
* [DENG-8210](https://mozilla-hub.atlassian.net/browse/DENG-8210)
* [HNT-413](https://mozilla-hub.atlassian.net/browse/HNT-413)

**Reviewer, please follow [this checklist](https://github.com/mozilla/bigquery-etl/blob/main/.github/reviewer_checklist.md)**

[DENG-8210]: https://mozilla-hub.atlassian.net/browse/DENG-8210?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[HNT-413]: https://mozilla-hub.atlassian.net/browse/HNT-413?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 1]
```diff
@@ -0,0 +1,112 @@
+WITH events_dedupe AS (
+  SELECT
+    *
+  FROM
+    `moz-fx-data-shared-prod.firefox_desktop_stable.newtab_v1`
```
Comment (kik-kik) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:5]: Shouldn't the data in `_stable` already be deduplicated as part of the `copy_deduplicate` run?

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    category AS event_category,
+    name AS event_name,
+    extra AS event_details,
+  FROM
+    events_dedupe,
+    UNNEST(events)
+  WHERE
+    category IN ('pocket')
+    AND name IN ('impression', 'click', 'save', 'dismiss')
+    AND mozfun.norm.browser_version_info(client_info.app_display_version).major_version >= 121
```
[the remark this answers is not in the document]
Comment (gkatre) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:28]: I believe it was the release when `pocket` events were added.
I picked it from the [code](https://github.com/Pocket/data-flows/blob/main-v2/data-products/src/firefox_newtab/sql/glean_firefox_new_tab_impressions_hourly/glean_firefox_new_tab_daily_engagement_by_scheduled_corpus_item_id_position_country_locale/data.sql#L49) that powered the original Mode reports in Snowflake

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+friendly_name: Newtab Clients Daily
```
Comment (chelseybeck) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml:1]:
```suggestion
friendly_name: Newtab Items Daily
```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/schema.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- name: locale
+  type: STRING
+  mode: NULLABLE
+- name: country
+  type: STRING
+  mode: NULLABLE
+- name: scheduled_surface_id
+  type: STRING
+  mode: NULLABLE
+- name: corpus_item_id
```
[the remark this answers is not in the document]
Comment (chelseybeck) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/schema.yaml:20]: We should have descriptions on every field...that's a requirement we have instituted :)

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql (hunk 1 above)
Comment (gkatre) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:5]: You are correct. I do not see any duplicates for `document_id`. I will remove the filter. Thanks!

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:19]: +1 to add a comment in this file directly, e.g.
```suggestion
    AND mozfun.norm.browser_version_info(client_info.app_display_version).major_version >= 121 -- the [Pocket team started using Glean](https://github.com/Pocket/dbt-snowflake/pull/459) from this version on. This prevents duplicates for previous releases.```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/schema.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/schema.yaml:25]: +1

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      AND (SPLIT(locale, '-')[1] IN ('GB', 'IE') OR country IN ('GB', 'IE'))
+      THEN 'NEW_TAB_EN_GB'
+    WHEN SPLIT(locale, '-')[0] = 'en'
+      AND (SPLIT(locale, '-')[1] = 'IN' OR country = 'IN')
+      THEN 'NEW_TAB_EN_INTL'
+    WHEN SPLIT(locale, '-')[0] = 'en'
+      AND (SPLIT(locale, '-')[1] IN ('US', 'CA') OR country IN ('US', 'CA'))
+      THEN 'NEW_TAB_EN_US'
+    ELSE 'NEW_TAB_EN_US'
+  END AS scheduled_surface_id,
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:70]: Should we move this classification to a UDF and document it to facilitate adding/removing and using it elsewhere consistently?

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  position,
+  is_sponsored,
+  is_secton_followed,
+  matches_selected_topic,
+  received_rank,
+  section,
+  section_position,
+  topic,
+  SUM(IF(event_name = 'impression', 1, 0)) AS impression_count,
+  SUM(IF(event_name = 'click', 1, 0)) AS click_count,
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:81]:
```suggestion
  COUNTIF(event_name = 'click'') AS click_count,
```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  SUM(IF(event_name = 'save', 1, 0)) AS save_count,
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:82]:
```suggestion
  COUNTIF(event_name = 'save') AS save_count,
```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  SUM(IF(event_name = 'dismiss', 1, 0)) AS dismiss_count,
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:83]:
```suggestion
  COUNTIF(event_name = 'save') AS save_count,
  COUNTIF(event_name = 'dismiss') AS dismiss_count,
```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    locale,
+    country,
+    event_category,
+    event_name,
+    mozfun.map.get_key(event_details, 'corpus_item_id') AS corpus_item_id,
+    SAFE_CAST(mozfun.map.get_key(event_details, 'position') AS INT64) AS position,
+    SAFE_CAST(mozfun.map.get_key(event_details, 'is_sponsored') AS BOOLEAN) AS is_sponsored,
+    SAFE_CAST(
+      mozfun.map.get_key(event_details, 'is_secton_followed') AS BOOLEAN
+    ) AS is_secton_followed,
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:35]:
```suggestion
    ) AS is_section_followed,
```

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+friendly_name: Newtab Items Daily
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml:1]: Would it be feasible to derive this table from new_tab_visits?

**Review by lucia-vargas-a — CHANGES_REQUESTED** [review]
Recommending a UDF, descriptions and code improvements. Curios whether this table would better be derived from new_tab_visits_v1.

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml (hunk 10 above)
Comment (gkatre) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml:1]: new_tab_visits currently does not have all the needed fields. We should eventually think of consolidating the  newtab etl processes. Perhaps as part of Nadia's revamp of the data models revisit all reporting requirements before we do it

Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/metadata.yaml:1]: Makes sense, it looks like we mostly derive from stable tables so we'd need to make this columns available.

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql [hunk 11]
```diff
@@ -0,0 +1,35 @@
+
+                -- The value returned by newtab.scheduled_surface_id_v1 represents the Content teams reference for the
+                -- surface where an article is published on the Newtab.
+                -- This UDF is composed based on the merino backend implementation here:
+                -- https://github.com/mozilla-services/merino-py/blob/main/merino/curated_recommendations/provider.py#L66-L103
```
Comment (lucia-vargas-a) [inline sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:5]:
```suggestion
-- The value returned by newtab.scheduled_surface_id_v1 represents the Content teams reference for the
-- surface where an article is published on the Newtab.
-- This UDF is composed based on the merino backend implementation here:
-- https://github.com/mozilla-services/merino-py/blob/main/merino/curated_recommendations/provider.py#L66-L103
```

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+      AND (SPLIT(locale, '-')[1] = 'IN' OR country = 'IN')
+      THEN 'NEW_TAB_EN_INTL'
+    WHEN SPLIT(locale, '-')[0] = 'en'
+      AND (SPLIT(locale, '-')[1] IN ('US', 'CA') OR country IN ('US', 'CA'))
+      THEN 'NEW_TAB_EN_US'
+    ELSE 'NEW_TAB_EN_US'
+  END
+);
+
+                -- Tests
```
Comment (lucia-vargas-a) [inline sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:30]:
```suggestion
-- Tests
```

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+SELECT
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('US', 'en-US')),
+  assert.equals('NEW_TAB_EN_GB', newtab.scheduled_surface_id_v1('IE', 'en-IE')),
+  assert.equals('NEW_TAB_EN_INTL', newtab.scheduled_surface_id_v1('IN', 'en-US')),
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('CA', 'en-US'));
```
Comment (lucia-vargas-a) [inline sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:35]: Is this correct? Wouldn't a combination of NULLs return NULL? E.g.

```suggestion
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('CA', 'en-US'));
  -- Any other country - locale combination will be classified as NEW_TAB_EN_US
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('DE', 'arch')),
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('DE', NULL))
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1(NULL, NULL))
```

**Review by lucia-vargas-a — CHANGES_REQUESTED** [review]
Shouldn't NULLs should return NULL in the UDF?

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql (hunk 13 above)
Comment (gkatre) [reply sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:35]: > Is this correct? Wouldn't a combination of NULLs return NULL?

I think the default they want to use is 'NEW_TAB_EN_US' if the country and locale are NULLs.

**Comment by gkatre** [comment]
> Shouldn't NULLs should return NULL in the UDF?

@lucia-vargas-a Thanks for looking at everything in detail!

I think the [default](https://github.com/mozilla-services/merino-py/blob/main/merino/curated_recommendations/provider.py#L92-L93) they want to use is `NEW_TAB_EN_US`.

I also think this UDF is going to be a placeholder until the newtab ping includes the field in the schema

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+-- Tests
+SELECT
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('US', 'en-US')),
+  assert.equals('NEW_TAB_EN_GB', newtab.scheduled_surface_id_v1('IE', 'en-IE')),
+  assert.equals('NEW_TAB_EN_INTL', newtab.scheduled_surface_id_v1('IN', 'en-US')),
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('CA', 'en-US')),
+  -- Any other country - locale combination of country-locale will be classified as NEW_TAB_EN_US, including NULL values.
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('DE', 'arch')),
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('DE', NULL)),
+  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1(NULL, NULL));
```
Comment (jpetto) [inline sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:39]: the logic looks to correctly match the referenced code in merino, but i would add a few more tests to ensure `locale` values like "de" and "en" work as expected. see the related tests in merino here: https://github.com/mozilla-services/merino-py/blob/main/tests/unit/curated_recommendations/test_provider.py#L70

Comment (mmiermans) [reply sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql:39]: I asked ChatGPT to find edge cases where the SQL would produce a different answer than the Python code. Here are the ones it came up with: 
```sql
-- 1. Case Sensitivity (Non-English language):
--    Locale "DE-DE" (uppercase) will not match 'de' due to case sensitivity,
--    so the SQL implementation falls through to the default.
SELECT
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('DE', 'DE-DE'));

-- 2. Locale Without a Dash:
--    A locale like "en" yields no region part, so the SQL check on SPLIT(locale, '-')[1]
--    fails, resulting in the default English US surface.
SELECT
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('US', 'en'));

-- 3. Inconsistent Input Parameters:
--    Here the locale suggests a GB region ("en-GB") but the country is provided as "US".
--    The SQL logic uses the country in its check, resulting in NEW_TAB_EN_US.
SELECT
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('US', 'en-GB'));

-- 4. Ordering / Inconsistent Region Edge Case:
--    With a locale "en" (no dash, so no region extracted) and the country provided as "GB",
--    the SQL logic falls back on the country check and returns NEW_TAB_EN_GB.
SELECT
  assert.equals('NEW_TAB_EN_GB', newtab.scheduled_surface_id_v1('GB', 'en'));

-- 5. Case Sensitivity for Non-English (French):
--    Locale "FR-FR" (uppercase language code) fails to match 'fr' in SQL,
--    so it falls to the default, returning NEW_TAB_EN_US rather than NEW_TAB_FR_FR.
SELECT
  assert.equals('NEW_TAB_EN_US', newtab.scheduled_surface_id_v1('FR', 'FR-FR'));
```

It says the comparison is case-sensitive, which doesn't sound right to me, but it's an edge case I would test regardless. The next two make some sense to me.

File: sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    mozfun.map.get_key(event_details, 'newtab_visit_id') AS newtab_visit_id,
+    SAFE_CAST(mozfun.map.get_key(event_details, 'received_rank') AS INT64) AS received_rank,
+    mozfun.map.get_key(event_details, 'section') AS section,
+    SAFE_CAST(mozfun.map.get_key(event_details, 'section_position') AS INT64) AS section_position,
+    mozfun.map.get_key(event_details, 'topic') AS topic,
+  FROM
+    events_unnested
+)
+SELECT
+  submission_date,
```
Comment (rolf-moz) [inline sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:48]: I think original snowflake processing did reporting for each surface based on the time zone of the surface. Are we going to continue that with this table, or will it be based on UTC time zone?

Comment (gkatre) [reply sql/moz-fx-data-shared-prod/firefox_desktop_derived/newtab_items_daily_v1/query.sql:48]: I would it makes sense to standardize the time zones across all existing reporting

File: sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql (hunk 14 above)
Comment (gkatre) [reply sql/mozfun/newtab/scheduled_surface_id_v1/udf.sql]: @jpetto @mmiermans  Thanks for the suggestions!

Based on the merino tests (those i have included) requires case-insensitive rules to be applied. 
I have updated the UDF accordingly and all the merino tests now seem to pass.

If i apply the ChatGPT edge cases, those will therefore fail
example: 
- country: 'DE', locale: 'DE-DE' will return 'NEW_TAB_DE_DE'.
- i have also set the precedence to country over locale. Therefore, country: 'US', locale: 'en-GB' will result in 'NEW_TAB_EN_US'

Let me know if I should make any changes.

File: sql/mozfun/newtab/scheduled_surface_id_v1/README.md [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+<!--
+This is a short README for your routine, you can add any extra
+documentation or examples that a user might want to see when
+viewing the documentation at https://mozilla.github.io/bigquery-etl
+
+You can embed an SQL file into your README using the following
+syntax:
+
+@sql(../examples/fenix_app_info.sql)
+-->
```
Comment (lucia-vargas-a) [inline sql/mozfun/newtab/scheduled_surface_id_v1/README.md:11]:
```suggestion
```

**Review by lucia-vargas-a — APPROVED** [review]
